drivers/openxr/openxr_device.h

```diff
@@ -0,0 +1,299 @@
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef OPENXR_DRIVER_H
+#define OPENXR_DRIVER_H
+
+#ifdef OPENXR_DUMMY
+// if OpenXR is not supported on the platform, we still have other parts of Godot
+// attempting to use the OpenXR driver, so we need a dummy
+
+class OpenXRDevice {
+public:
+	static void setup_global_defs(){};
+	static bool openxr_is_enabled() { return false; };
+	static OpenXRDevice *get_singleton() { return nullptr; };
+
+	bool is_initialized() { return false; };
+	bool is_running() { return false; };
+	bool initialise(const String &p_rendering_driver) { return false; };
+	bool initialise_session() { return false; };
+	void finish(){};
+};
+
+#else
+
+#include "core/error/error_macros.h"
+#include "core/math/camera_matrix.h"
+#include "core/math/transform_3d.h"
+#include "core/math/vector2.h"
+#include "core/os/memory.h"
+#include "core/string/ustring.h"
+#include "core/templates/map.h"
+#include "core/templates/rid_owner.h"
+#include "core/templates/vector.h"
+
+#include "thirdparty/openxr/src/common/xr_linear.h"
+#include <openxr/openxr.h>
+
+#include "openxr_composition_layer_provider.h"
+#include "openxr_extension_wrapper.h"
+
+// Note, OpenXR code that we wrote for our plugin makes use of C++20 notation for initialising structs which ensures zeroing out unspecified members.
+// Godot is currently restricted to C++17 which doesn't allow this notation. Make sure critical fields are set.
```

Comment:
@BastiaanOlij @akien-mga Is it possible to leverage `clang-format` to ensure this is properly observed?

Reply:
`clang-format` is only a formatting tool, it can't change code.

`clang-tidy` can do code analysis and changes, including giving a default initialization to uninitialized variables, though it can be somewhat unwieldy to use (applies to the whole case, has false positives and creates compilation bugs).

But what's this C++20 notation that we can't use here? If it's `{}`, AFAIK it's available in C++17, we use it in some places.

Reply:
@akien-mga this does not work on C++17, atleast for MSVC:
```
	XrEventDataBuffer runtimeEvent = {
		.type = XR_TYPE_EVENT_DATA_BUFFER,
		.next = nullptr
	};
```

And even on C++20 MSVC has small syntax differences compared to LLVM and MSVC requires you to keep the fields in the order of the struct, while on other compilers you can reorder them. On both though you can leave fields out and they will be zero filled.

That said, this syntax is far more readable and far easier to understand than the C++17 syntax I'm currently using (though the comments help in accomplishing the same :))

Reply:
I think in C++17 you can do:
```
XrEventDataBuffer runtimeEvent = {};
// or XrEventDataBuffer runtimeEvent {};
runtimeEvent.type = XR_TYPE_EVENT_DATA_BUFFER;
```
The init with `{}` should zero everything in the struct and then you can set the members you care about outside the init.

Reply:
Tested, works as early as C++11:
```cpp
#include <iostream>

class SomeClass {
public:
	int class_a = 5;
	SomeClass() {}
};

struct SomeStruct {
	int some_int;
	float some_float;
	SomeClass some_obj;
	SomeClass *some_ptr;
};

int main() {
	SomeStruct some_str {};
	std::cout << some_str.some_int << " " << some_str.some_float << " " << some_str.some_obj.class_a << " " << some_str.some_ptr << std::endl;
	SomeStruct other_str;
	std::cout << other_str.some_int << " " << other_str.some_float << " " << other_str.some_obj.class_a << " " << other_str.some_ptr << std::endl;
	return 0;
}
```
```
$ g++ -std=c++11 test.cpp
$ ./a.out
0 0 5 0
-327758120 4.55576e-41 5 0x7effec76d160
```

Reply:
> The init with `{}` should zero everything in the struct and then you can set the members you care about outside the init.
We found out the hard way (note Godot crashing hard), that this is not true on every platform. Android did not zero out things when initialized this way and it ended up crashing stuff when the next pointer had a random value instead of being zero.